Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -692,104 +688,75 @@ public class ProgramBuilder {
+                slice.insert(instr.index)
             }
-            requiredInputs.formUnion(neededInputs)
-            requiredInstructions.insert(instr.index)
-        }
 
-        func keep(_ instr: Instruction, includeBlockContent: Bool = false) {
-            guard !requiredInstructions.contains(instr.index) else { return }
             if instr.isBlock {
                 let group = BlockGroup(around: instr, in: source)
-                let instructions = includeBlockContent ? group.includingContent() : group.excludingContent()
-                for instr in instructions {
-                    rewireOrKeepInputs(of: instr)
+                for instr in includeBlockContent ? group.includingContent() : group.excludingContent() {
+                    internalAdd(instr)
                 }
             } else {
-                rewireOrKeepInputs(of: instr)
+                internalAdd(instr)
             }
         }
 
-        // Keep the selected instruction
-        keep(program.code[idx], includeBlockContent: true)
+        // Compute the slice...
+        var idx = index
+        
+        // First, add the selected instruction.
+        add(source[idx], includeBlockContent: true)
 
+        // Then add all instructions that the slice has data dependencies on.
         while idx > 0 {
             idx -= 1
-            let current = source[idx]
+            let instr = source[idx]
 
-            if !requiredInputs.isDisjoint(with: current.allOutputs) {
-                let onlyNeedsInnerOutputs = requiredInputs.isDisjoint(with: current.outputs)
+            if !requiredInputs.isDisjoint(with: instr.allOutputs) {
+                let onlyNeedsInnerOutputs = requiredInputs.isDisjoint(with: instr.outputs)
                 // If we only need inner outputs (e.g. function parameters), then we don't include
                 // the block's content in the slice. Otherwise we do.
-                keep(current, includeBlockContent: !onlyNeedsInnerOutputs)
+                add(instr, includeBlockContent: !onlyNeedsInnerOutputs)
             }
 
             // If we perform a potentially mutating operation (such as a property store or a method call)
             // on a required variable, then we may decide to keep that instruction as well.
             if mode == .conservative || (mode == .aggressive && probability(0.5)) {
-                if current.mayMutate(requiredInputs) {
-                    keep(current, includeBlockContent: false)
+                if instr.mayMutate(requiredInputs) {
+                    add(instr)
                 }
             }
         }
 
+        // Finally, insert the slice into the current program.
         for instr in source {
-            if requiredInstructions.contains(instr.index) {
+            if slice.contains(instr.index) {
                 adopt(instr, keepTypes: true)
             }
         }
 
         endAdoption()
-        trace("End of splice")
+        trace("Splicing done")
```

Comment:
Would it make sense to add an assertion here to check if code is statically valid? Not sure if code should always be statically valid at the end of a splice operation.

Reply:
No I think that wouldn't work, since you might be splicing after e.g. a `StartBlockX` operation but before the `EndBlockX`, in which case the _current_ code wouldn't be statically valid (but the final code will be).